File: unique binary search tree ii.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def dfs(self, start, end):
        if start > end:         #如果ｓｔａｒｔ比ｅｎｄ大说明这棵树的所有节点都已经被遍历过　返回[None]表示下面没有节点了
            return [None]
        tmp = []               #每次用一个ｌｉｓｔ　ｔｍｐ来存放当前树的根节点
        for rootval in range(start, end+1):         #循环对每一个节点都执行一次操作
            leftSub = self.dfs(start, rootval-1)    #递归建立当前根节点的左子树　ｂｉｎａｒｙ　ｓｅａｒｃｈ　ｔｒｅｅ的左小右大的逻辑在此建立(start, rootval-1), (rootval+1, end)
            rightSub = self.dfs(rootval+1, end)     #递归建立当前根节点的右子树
            for i in leftSub:           #建立根节点　添加左右子树　并将根节点ａｐｐｅｎｄ到ｔｍｐ　ｌｉｓｔ中去
                for j in rightSub:
                    root = TreeNode(rootval)
                    root.left = i
                    root.right = j
                    tmp.append(root)
        return tmp      #最后返回ｔｍｐ
    
    # @return a list of tree node
    def generateTrees(self, n):
        return self.dfs(1, n)

Note: 这题也要看草稿图　以ｎ=2为例　看下执行过程　比较好理解





题意：接上一题，这题要求返回的是所有符合条件的二叉查找树，而上一题要求的是符合条件的二叉查找树的棵数，我们上一题提过，求个数一般思路是动态规划，而枚举的话，我们就考虑dfs了。dfs(start, end)函数返回以start，start+1，...，end为根的二叉查找树。

代码：


# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @return a list of tree node
    def dfs(self, start, end):
        if start > end: return [None]
        res = []
        for rootval in range(start, end+1):　　　　　　　　#rootval为根节点的值，从start遍历到end
            LeftTree = self.dfs(start, rootval-1)
            RightTree = self.dfs(rootval+1, end)
            for i in LeftTree:　　　　　　　　　　　　　　　　#i遍历符合条件的左子树
                for j in RightTree:　　　　　　　　　　　　  #j遍历符合条件的右子树
                    root = TreeNode(rootval)
                    root.left = i
                    root.right = j
                    res.append(root)
        return res
    def generateTrees(self, n):
        return self.dfs(1, n)

File: unique binary search tree ii.java

public ArrayList<TreeNode> generateTrees(int n) {
    return helper(1,n);
}
private ArrayList<TreeNode> helper(int left, int right)
{
    ArrayList<TreeNode> res = new ArrayList<TreeNode>();
    if(left>right)
    {
        res.add(null);
        return res;
    }
    for(int i=left;i<=right;i++)            //遍历 分别以1...n为根节点
    {
        ArrayList<TreeNode> leftList = helper(left,i-1);    //递归建立根节点的左右子树 helper返回的是以i为根 所有可能的子树根节点
        ArrayList<TreeNode> rightList = helper(i+1,right);
        for(int j=0;j<leftList.size();j++)          //遍历leftList rightList里的节点 每一个节点代表当前根(root)左或右子树的根节点 子树下面的结构在之前的递归中已经建立好
        {
            for(int k=0;k<rightList.size();k++)
            {
                TreeNode root = new TreeNode(i);       //这里 建立树根 直接将leftList rightList里的节点连接在root上即可 节点下面的子树结构已在之前的递归中建立好了 
                root.left = leftList.get(j);
                root.right = rightList.get(k);
                res.add(root);                  //最后将这个root（一个root具有一种特定子树结构 同一个root可以有多种子树结构）加到res中
            }
        }
    }
    return res;
}

Note: 看generateTree.java generateTreeTest.java程序理解 里面有打印信息

这题想了半天才明白 着重看下注释 还可以跑下相关的generateTreeTest.java程序 看输出理解




这道题是求解所有可行的二叉查找树，从Unique Binary Search Trees中我们已经知道，可行的二叉查找树的数量是相应的卡特兰数，

不是一个多项式时间的数量级，所以我们要求解所有的树，自然是不能多项式时间内完成的了。算法上还是用求解NP问题的方法来求解，

也就是N-Queens中介绍的在循环中调用递归函数求解子问题。思路是每次一次选取一个结点为根，然后递归求解左右子树的所有结果，

最后根据左右子树的返回的所有子树，依次选取然后接上（每个左边的子树跟所有右边的子树匹配，而每个右边的子树也要跟所有的左边子树匹配，

总共有左右子树数量的乘积种情况），构造好之后作为当前树的结果返回。代码如下： 

public ArrayList<TreeNode> generateTrees(int n) {
    return helper(1,n);
}
private ArrayList<TreeNode> helper(int left, int right)
{
    ArrayList<TreeNode> res = new ArrayList<TreeNode>();
    if(left>right)
    {
        res.add(null);
        return res;
    }
    for(int i=left;i<=right;i++)
    {
        ArrayList<TreeNode> leftList = helper(left,i-1);
        ArrayList<TreeNode> rightList = helper(i+1,right);
        for(int j=0;j<leftList.size();j++)
        {
            for(int k=0;k<rightList.size();k++)
            {
                TreeNode root = new TreeNode(i);
                root.left = leftList.get(j);
                root.right = rightList.get(k);
                res.add(root);
            }
        }
    }
    return res;
}

实现中还是有一些细节的，因为构造树时两边要遍历所有左右的匹配，然后接到根上面。

当然我们也可以像在Unique Binary Search Trees中那样存储所有的子树历史信息，然后进行拼合，虽然可以省一些时间，

但是最终还是逃不过每个结果要一次运算，时间复杂度还是非多项式的，并且要耗费大量的空间，感觉这样的意义就不是很大了。

